### PR TITLE
feat: allow passing assets instead of `bytes` for image parameters

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

A QoL improvement that changes all methods (except `SyncWebhook.edit`) which previously accepted some kind of image as bytes to also transparently handle `AssetMixin` objects, which includes things like user/role/guild icons, emojis, stickers, etc.

```py
# before
await role.edit(icon=await member.avatar.read())

# after
await role.edit(icon=member.avatar)
```

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: disnake/guild.py [hunk 1]
```diff
@@ -1644,6 +1661,8 @@ async def edit(
 
         Raises
         ------
+        NotFound
+            One of the assets (``icon``, ``banner``, ``splash`` or ``discovery_splash``) couldn't be found.
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/guild.py]: d03e728feb8a960b9f7b9553c58347d1393f3a23

File: disnake/asset.py [hunk 2]
```diff
@@ -34,12 +34,17 @@
 from . import utils
 from .errors import DiscordException, InvalidArgument
 
-__all__ = ("Asset",)
+__all__ = (
+    "AssetMixin",
```
Comment (onerandomusername) [inline disnake/asset.py:38]: this shouldn't be here

File: disnake/asset.py [hunk 3]
```diff
@@ -34,12 +34,17 @@
...
+    "Asset",
+)
 
 if TYPE_CHECKING:
     ValidStaticFormatTypes = Literal["webp", "jpeg", "jpg", "png"]
     ValidAssetFormatTypes = Literal["webp", "jpeg", "jpg", "png", "gif"]
 
+    AssetBytes = Union[bytes, "AssetMixin"]
```
Comment (onerandomusername) [inline disnake/asset.py:46]: this shouldn't be declared in TYPE_CHECKING if we're importing it into other files IMO

Comment (onerandomusername) [reply disnake/asset.py:46]: We've had this problem before, with #384 iirc

File: disnake/asset.py [hunk 4]
```diff
@@ -48,6 +53,17 @@
...
+    """
+    Base type for all asset-like types.
+    This currently includes:
+
+    - :class:`Asset`
+    - :class:`Emoji`
+    - :class:`PartialEmoji`
+    - :class:`StickerItem`
+    - :class:`Sticker`
+    """
```
Comment (onerandomusername) [inline disnake/asset.py:65]: probably can be removed? Not entirely sure

Comment (shiftinv) [reply disnake/asset.py]: cc3f3fc22552f4f439545ea6d7997b8a3725709d

File: disnake/asset.py (hunk 2 above)
Comment (shiftinv) [reply disnake/asset.py]: cc3f3fc22552f4f439545ea6d7997b8a3725709d

File: disnake/asset.py (hunk 3 above)
Comment (shiftinv) [reply disnake/asset.py]: 79364942bc96ce3814f3a221dd1de77c65eb001e

**Comment by onerandomusername** [comment]
Small problem discovered which relates to stickers. Looking into it now.

**Comment by onerandomusername** [comment]
Could you add a note about Stickers, or perhaps remove them from the docs entirely being officially supported to be passed directly? 

Lottie stickers through the whole thing off here. This is also an issue with #415 (which was just merged).
https://github.com/DisnakeDev/disnake/blob/a6d749c55eb633c745702de32c0bbc565309dc5a/disnake/sticker.py#L169-L170

**Comment by onerandomusername** [comment]
We should add TypeError to the docs of anything that can take an asset directly, as certain Sticker types will raise said exception.

**Comment by shiftinv** [comment]
> We should add TypeError to the docs of anything that can take an asset directly, as certain Sticker types will raise said exception.

https://github.com/DisnakeDev/disnake/pull/415/commits/d52e316a65f39bcc001e1c02a6c0b0b44a93dcc5

**Review by onerandomusername — APPROVED** [review]
looks good

**Pull request merged**